pkg/target/http.go

```diff
@@ -226,52 +268,175 @@ func (ht *HTTPTarget) Write(messages []*models.Message) (*models.TargetWriteResu
+	//ACK and set healthy on success...
+	if ht.isSuccess(response) {
+		health.SetHealthy()
+		sent = append(sent, msg)
+		if msg.AckFunc != nil {
+			msg.AckFunc()
 		}
-		defer resp.Body.Close()
-		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
-			sent = append(sent, msg)
-			if msg.AckFunc != nil { // Ack successful messages
-				msg.AckFunc()
-			}
-		} else {
-			errResult = multierror.Append(errResult, errors.New("Got response status: "+resp.Status))
-			failed = append(failed, msg)
-			continue
+		return nil
+	}
+
+	if ht.isInvalid(response) {
+		// App is healthy, our data is not
+		health.SetHealthy()
+		invalid = append(invalid, msg)
+		return nil
+	}
+
+	if currentRule := ht.findRetryableRule(response); currentRule != nil {
+		//we found matching rule! Start retrying
+		return ht.handleRetryableResponse(currentRule, response, previouslyMatchedRule, request, msg, sent, invalid, failed, errResult)
+	}
+
+	//no success/invalid/retryable rule matches, so we fallback to 'old default' layer of retrying (in main) and append to failed list.
+	errResult = multierror.Append(errResult, errors.New("TODO"))
+	failed = append(failed, msg)
+	return nil
+}
+
+func (ht *HTTPTarget) handleRetryableResponse(currentRule *Rule, response Response, previouslyMatchedRule *Rule, request *http.Request, msg *models.Message, sent []*models.Message, invalid []*models.Message, failed []*models.Message, errResult error) error {
+	health.SetUnhealthy()
+	//send alert if configured...
+	if currentRule.Alert != "" {
+		ht.monitoring.SendAlert(monitoring.Alert{Message: fmt.Sprintf("%s, response body - %s", currentRule.Alert, response.Body)})
+	}
+	// If we have some 'previouslyMatchedRule' then we know we're currently retrying some request. It answers question - 'are we currently retrying something??'.
+	// If it's equal to the new rule we've just matched, we have to retry again using the same strategy, instead of starting new retry loop. That's why we stop here and return control to the caller.
+	if currentRule == previouslyMatchedRule {
+		return errors.New("Same rule, probably same error, retry using the same strategy in upper layer")
+	}
+
+	// If `previouslyMatchedRule` is not defined (e.g. at the beginning it's ) or is not equal to the new rule (so we have some different error to deal with) we change retrying strategy.
+	strategy := ht.responseHandler.RetryStrategies[currentRule.RetryStrategy]
+
+	ht.retry(strategy, func() (interface{}, error) {
+		response := ht.executeHTTPRequest(request, msg)
+		return nil, ht.handleHTTPResponse(request, response, msg, sent, invalid, failed, errResult, currentRule)
+	})
+	return nil
+}
+
+func (ht *HTTPTarget) retry(config RetryConfig, action func() (interface{}, error)) interface{} {
+	// Should we use some third-party retrying library? Where we can configure more stuff.
```

Comment:
If we find a well supported library that does it, let's look at that - especially if it can simplify our implementation.

Reply:
Also have the option of adding what we need to this package